### ISSUE TITLE
remote forward extreamly slow when ray transmits large data batch

### ISSUE THREAD

**Issue by TideDra** [OpenRLHF/OpenRLHF#998]
The single controller has a serious communication bottleneck and memory leakage when setting a large batchsize=256, n_samples_per_prompt=16 and max_len=8192

When making experience, the controller sends all the data to the actors at one time. This communication cost seems to increase non-linearly with the data size. For the default setting in the example script (bs=128, n_samples_per_prompt=8), it takes seconds. While for bs=256, n=16, it takes >10 minutes and even timeout. The memory consumption also increase unnormally, trigerring ray spilling objects.
<img width="862" alt="Image" src="https://github.com/user-attachments/assets/c3281cd0-227c-48bc-aae5-d4799e97ee64" />

<img width="1357" alt="Image" src="https://github.com/user-attachments/assets/7c456e2b-943a-4f87-aa19-4d8a58df6422" />

My script:

```bash
ray job submit --address="http://127.0.0.1:8265" \
   --runtime-env-json='{"working_dir": "./"}' \
   -- python3 -m openrlhf.cli.train_ppo_ray \
   --ref_num_nodes 1 \
   --ref_num_gpus_per_node 8 \
   --reward_num_nodes 1 \
   --reward_num_gpus_per_node 8 \
   --actor_num_nodes 1 \
   --actor_num_gpus_per_node 8 \
   --vllm_num_engines 4 \
   --vllm_tensor_parallel_size 2 \
   --colocate_all_models \
   --vllm_gpu_memory_utilization 0.6 \
   --advantage_estimator reinforce_baseline \
   --pretrain /root/ckpts/DeepSeek-R1-Distill-Qwen-1.5B \
   --reward_pretrain /root/ckpts/DeepSeek-R1-Distill-Qwen-1.5B \
   --save_path /root/ckpts/test \
   --ckpt_path /root/ckpts/test \
   --save_hf_ckpt \
   --micro_train_batch_size 4 \
   --train_batch_size 128 \
   --micro_rollout_batch_size 8 \
   --rollout_batch_size 256 \
   --n_samples_per_prompt 16 \
   --init_kl_coef 1e-3 \
   --gamma 1.0 \
   --use_kl_loss \
   --kl_estimator k2 \
   --max_epochs 1 \
   --prompt_max_len 1024 \
   --max_samples 100000 \
   --generate_max_len 8192 \
   --zero_stage 3 \
   --bf16 \
   --actor_learning_rate 5e-7 \
   --critic_learning_rate 9e-6 \
   --prompt_data agentica-org/DeepScaleR-Preview-Dataset \
   --input_key problem \
   --apply_chat_template \
   --normalize_reward \
   --gradient_checkpointing \
   --packing_samples \
   --ring_attn_size 2 \
   --ring_head_stride 2 \
   --vllm_sync_backend nccl \
   --vllm_enable_sleep \
   --deepspeed_enable_sleep
```

**Comment by hijkzzz** [comment]
try https://github.com/OpenRLHF/OpenRLHF/pull/996/files
I optimized the ray transmits

**Comment by TideDra** [comment]
It doesn't work, ray.put still receives a large batch and then stucks.

**Comment by TideDra** [comment]
same issue reported in lmm-r1: https://github.com/TideDra/lmm-r1/issues/71
I tried to split the rollout_samples and call `make_experiences` multiple times to avoid transmitting large batch, but it's still slow. Maybe there is memory leakage in somewhere else

**Comment by hijkzzz** [comment]
It might not be a memory leak; the samples padding could be too large. We may need to optimize from this perspective.
that's just the problem with a single controller...........

**Comment by hijkzzz** [comment]
fixed in https://github.com/OpenRLHF/OpenRLHF/commit/ccb58df8102d08440d05efe24c08219bb2193756
